### PR TITLE
chore(STONEO11Y-97): Alert rule for degraded argoCD apps

### PR DESCRIPTION
**Pull request by ShakedAviv50313**
* Create an alerting rule for argoCD apps with 'Degraded' health status in the past 5 minutes.

### CODE REVIEW

File: test/promql/tests/test_degraded_argoCD.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      - eval_time: 15m
+        alertname: DegradedArgoCDApp
+        exp_alerts:
+          - exp_labels:
+              severity: Critical
+              dest_namespace: has-degraded
+              health_status: Degraded
+            exp_annotations:
+              summary: "ArgoCD service at has-degraded namespace includes an app which was 'Degraded'"
+              description: "ArgoCD service at has-degraded namespace includes an app which was 'Degraded' in the last 5 minutes"
```
Comment (Omeramsc) [inline test/promql/tests/test_degraded_argoCD.yaml:27]: where do we see the last-5-min in the test?

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  - name: argoCD_alerts
+    rules:
+    - alert: DegradedArgoCDApp
+      expr: max_over_time(argocd_app_info{health_status="Degraded", dest_namespace!~".+-tenant"}[5m]) == 1
+      for: 15m
+      labels:
+        severity: Critical
+      annotations:
+        summary: "ArgoCD service at {{ $labels.dest_namespace }} namespace includes an app which was 'Degraded'"
+        description: "ArgoCD service at {{ $labels.dest_namespace }} namespace includes an app which was 'Degraded' in the last 5 minutes"
```
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.argoCD_alerts.yaml:19]: You can get the name of the application from the metric and produce more precise summary and descriptions.  
Maybe something like:
`description: "Application {{ $labels.name }} in namespace {{ $labels.dest_namespace }} was 'Degraded' in the last 5 minutes"`

Comment (Omeramsc) [inline prometheus/base/alerting/prometheus.argoCD_alerts.yaml:19]: Maybe we can rephrase it to look someway similar to:
"An app with a `Degraded` status was found in the last 5 minutes under ArgoCD's service at .... "
?

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  labels:
+    openshift.io/prometheus-rule-evaluation-scope: leaf-prometheus
+  name: o11y-argoCD-alerting-rules
+  namespace: o11y
+spec:
+  groups:
+  - name: argoCD_alerts
+    rules:
+    - alert: DegradedArgoCDApp
+      expr: max_over_time(argocd_app_info{health_status="Degraded", dest_namespace!~".+-tenant"}[5m]) == 1
```
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.argoCD_alerts.yaml:13]: If the application was degraded and then got fixed the output would look like:
1 1 0 0 0.
Using max_over_time for it will return true since the max value in this time period is 1, but the application was not degraded for the past 5 minutes (only for the first 2 minutes).  
Am I right?

File: test/promql/tests/test_degraded_argoCD.yaml (hunk 1 above)
Comment (ShakedAviv50313) [reply test/promql/tests/test_degraded_argoCD.yaml]: I added it just now, it is on lines 18-25.

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml (hunk 2 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml]: how about now?

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml (hunk 3 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml:13]: as for my understanding of the task, the rule should alert if an app's status was changed to "Degraded" in the last5 minutes. However, there's a chance I misunderstood the task.
@yftacherzog What do you think? maybe that's why it was "last_over_time" instead of "max_over_time"?

File: test/promql/tests/test_degraded_argoCD.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  - interval: 1m
+    input_series:
+
+      # no argoCD app is degraded
+      - series: 'argocd_app_info{health_status="Degraded", name="first-test-app", dest_namespace="not-degraded"}'
+        values: '0x15'
+
+      # argoCD has a degraded app
+      - series: 'argocd_app_info{health_status="Degraded", name="second-test-app", dest_namespace="has-degraded"}'
+        values: '1x15'
```
Comment (amisstea) [inline test/promql/tests/test_degraded_argoCD.yaml:16]: It'd be nice to either have a more complex series here or some additional tests to cover patterns like flapping.

File: test/promql/tests/test_degraded_argoCD.yaml
Comment (amisstea) [inline test/promql/tests/test_degraded_argoCD.yaml:1]: `argoCD` -> `argocd`

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml
Comment (amisstea) [inline prometheus/base/alerting/prometheus.argoCD_alerts.yaml:1]: `argoCD` -> `argocd`

**Review by amisstea — CHANGES_REQUESTED** [review]

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml (hunk 3 above)
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml:13]: @ShakedAviv50313 when we looked at real data together, I don't think we ever saw a value of zero. It was only one or none. With that in mind, I think we need at least some of the test cases to represent that. If I understand correctly, none means the value is missing, which in promtool is represented by an underscore. 

See this link: https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series

Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml:13]: I think `max_over_time` makes more sense here.

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      for: 15m
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.argoCD_alerts.yaml:14]: I'm trying to wrap my head around what will be the outcome of the combination of max over the last 5 minutes together with for the last 15 minutes.

I think it means that it will only be alerting if the status was degraded at some point at every 5 minutes window within the last 15 minutes. It's a bit convoluted for me. Is this what we wanted? Can we make it any simpler?

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml:14]: so you suggest we change that to also 5 minutes?

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml (hunk 3 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml]: Sure, it is now done and I added a new test case representing a flappy app with also None "_" values

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    openshift.io/prometheus-rule-evaluation-scope: leaf-prometheus
+  name: o11y-argoCD-alerting-rules
```
[the remark this answers is not in the document]
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml]: Done

File: prometheus/base/alerting/kustomization.yaml [hunk 7]
```diff
@@ -3,3 +3,4 @@ kind: Kustomization
 resources:
 - prometheus.pod_alerts.yaml
 - prometheus.quota_alerts.yaml
+- prometheus.argoCD_alerts.yaml
```
[the remark this answers is not in the document]
Comment (ShakedAviv50313) [reply prometheus/base/alerting/kustomization.yaml]: Done, changed to "argocd"

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  namespace: o11y
+spec:
+  groups:
+  - name: argoCD_alerts
```
[the remark this answers is not in the document]
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml]: Done

File: test/promql/tests/test_degraded_argoCD.yaml (hunk 4 above)
Comment (ShakedAviv50313) [reply test/promql/tests/test_degraded_argoCD.yaml]: Done, I added another test case for a flappy app

File: prometheus/base/alerting/prometheus.argoCD_alerts.yaml (hunk 5 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argoCD_alerts.yaml]: It is now 5m

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  labels:
+    openshift.io/prometheus-rule-evaluation-scope: leaf-prometheus
+  name: o11y-argocd-alerting-rules
+  namespace: o11y
+spec:
+  groups:
+  - name: argocd_alerts
+    rules:
+    - alert: DegradedArgocdApp
+      expr: max_over_time(argocd_app_info{health_status="Degraded", dest_namespace!~".+-tenant"}[5m]) == 1
```
Comment (amisstea) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:13]: One of the acceptance criteria is to ensure the degraded component's `service` label is included in the alert. This metric always returns the value of `gitops-service-argocd-metrics` no matter which `dest_namespace` is affected. Have you tried to solve this?

Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:13]: the summary and the description return the affected app's name

Comment (amisstea) [reply prometheus/base/alerting/prometheus.argocd_alerts.yaml:13]: The `name` label looks something like `gitopsdepl-0c63a1fd-fbd3-41d4-ba02-50290ad8681e` which isn't very descriptive and probably not useful when we eventually need to route this alert. The `service` label is consistently available in the other alert rules so it'd be nice to conform to that standard, if possible.

File: prometheus/base/alerting/prometheus.argocd_alerts.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+      for: 5m
+      labels:
+        severity: Critical
+      annotations:
+        summary: "Application {{ $labels.name }} in namespace {{ $labels.dest_namespace }} was 'Degraded'"
```
Comment (amisstea) [inline prometheus/base/alerting/prometheus.argocd_alerts.yaml:18]: `$labels.name` seems to be the name of the gitops deployment (see my other comment) rather than the application name.

**Review by amisstea — APPROVED** [review]
LGTM :+1:

**Review by hmariset — APPROVED** [review]
LGTM

**Comment by ShakedAviv50313** [comment]
The original commit had conflicts because my fork was not synced with the original o11y repo,
I tried fixing that and only made things worse, I'll continue working on this tomorrow, sorry.

File: test/promql/tests/test_degraded_argocd.yaml
Comment (yftacherzog) [inline test/promql/tests/test_degraded_argocd.yaml:1]: Can we add a time series with some other health status with value 1 to see that it does not alert for it?

Comment (ShakedAviv50313) [reply test/promql/tests/test_degraded_argocd.yaml:1]: How about now?

Comment (yftacherzog) [reply test/promql/tests/test_degraded_argocd.yaml:1]: I want to see that we don't alert if we have a time series of some other health_status with value 1. Something like this:

```

      - series: 'argocd_app_info{health_status="not-Degraded", name="healthy-app", dest_namespace="nothing-wrong"}'
        values: 1x15

Comment (ShakedAviv50313) [reply test/promql/tests/test_degraded_argocd.yaml:1]: Ok now I added a "Progressing" health state and the rule does not alert it (as expected)

**Review by yftacherzog — APPROVED** [review]

File: test/promql/tests/test_degraded_argoCD.yaml (hunk 1 above)
Comment (ShakedAviv50313) [reply test/promql/tests/test_degraded_argoCD.yaml]: Is it ok now?

Comment (Omeramsc) [reply test/promql/tests/test_degraded_argoCD.yaml]: LGTM
